[ROCm] ROCm-specific GPU utility functions.

whchung opened the pull request:
Common utility helpers between CUDa and ROCm are renamed:
- core/util/cuda_device_functions.h -> core/util/gpu_device_functions.h
- core/util/cuda_kernel_helper.h -> core/util/gpu_kernel_helper.h
- core/util/cuda_kernel_helper_test.cu.cc -> core/util/gpu_kernel_helper_test.cu.cc
- core/util/cuda_launch_config.h -> core/util/gpu_launch_cnofig.h

ROCm-specific changes for GPU device functions are also introduced in this PR.

GPU kernel invocation logic would be introduced in the next PR. All custom CUDA kernels are still launched as `CudaLaunchKernel` for now.

smit-hinsu reviewed (dismissed):
If we go with the rename then the tensorrt change looks good to me.

whchung commented:
@smit-hinsu Thanks for approving this PR. Failed CI targets are:
- MacOS Contrib
- MacOS Python2 and CC
- Windows Bazel
- Windows Bazel GPU

From the logs they seem to have nothing to do with this PR, but a configuration issue in CI test infrastructure. I'm wondering if it's possible to manually apply `ready-to-pull` label?

ymodak commented:
@whchung Can you please resolve the conflict? Thanks!

ymodak commented:
@whchung Can you please resolve the conflict? Thanks!

rthadur commented:
> @whchung Can you please resolve the conflict? Thanks!

Gentle ping

whchung commented:
@rthadur could you give another go for this PR? thanks. the failures on "Ubuntu Python3 PIP" and "Windows Bazel" don't seem to be related to this PR.

whchung commented:
ping?

rthadur commented:
@whchung request to re base your branch

whchung commented:
@rthadur sorry for being late. I've rebased the PR.

smit-hinsu reviewed (dismissed):
I will let @qqfish review this as he was working on this stuff earlier.

whchung commented:
@smit-hinsu , @qqfish ping?

whchung commented:
invite @yifef to this PR as it's closely related to #24295 which was just merged

dagamayank commented:
Correctly tagging per @whchung comment. /cc @yifeif :)

whchung commented:
Assigning the PR to @tatianashp and @thirupalanisamy as this PR is now blocking upstreaming some TensorFlow operators for ROCm.

chsigg commented:
> In subsequent PRs, ROCm-specific changes for GPU device functions and GPU kernel invocation logic / macro would be introduced.

Wen-Heng, would you mind pointing me to the PRs/code that is going to come after this one. It doesn't need to be complete, just give me an idea what's going to come. As it stands, renaming the files in this PR looks like unnecessary churn because they still provide CUDA-only functionality. I assume it makes more sense once everything is merged, and I would like to get idea what that will look like.

whchung commented:
@chsigg thanks. I’m working on a PR to demonstrate this. It’s taking a little more that I wanted as recently TF tip has migrated from <<<>>> to CudaLaunchKernel so our downstream fork needs to be changed a bit. I’ll manage to get a PR submitted today. In fact I think I might just amend this PR.

whchung commented:
@chsigg I've revised this PR to put (many) more meats in. Macros / functions names were changed to be more vendor-neutral, and some ROCm-specific device function implementation were added.

I haven't changed kernel launching logic `CudaLaunchKernel` in this PR yet. I'd like to submit it as another PR, as this PR has already become pretty big.

whchung commented:
@chsigg I've fixed all the issues from CI tests. Some failures in `Linux GPU` test target don't seem to be related to this PR. Would you mind take another look at this PR? Thanks!

whchung commented:
@chsigg ping?

chsigg commented:
I'm still not sold that renaming these files and function names provides real value. For example 
> `data += GpuShuffleXorSync(kCudaWarpAll, data, delta);`
is not HIP-ready, so it might be better to keep the original CUDA name.

Would you be ok with getting the _functionality_ in first, i.e. the #ifdef blocks to compute launch configs? The fact that the file and function names contain CUDA seems a cosmetic detail that we can overlook for now.

whchung commented:
@chsigg all the implementation details to get ROCm path functional have already been added inside this PR:

- use device intrinsics on AMDGPU to compute shuffles
- introduce GPU_LAUNCH_KERNEL to cope with differences in kernel launch syntax between CUDA and ROCm
- use #if blocks to distinguish differences between ROCm and CUDA runtime API calls, especially mark those places yet supported by HIP runtime APIs.

- and finally terminology changes (Cuda -> GPU)

I’m wondering could you do another round of review, especially pay attention to those files in the bottommost of this PR where most of these functionality changes take place?

chsigg commented on tensorflow/core/util/gpu_kernel_helper.h:
```diff
@@ -13,34 +13,76 @@ See the License for the specific language governing permissions and
+#endif
+
+#if GOOGLE_CUDA
+#define GPU_LAUNCH_KERNEL(kernel, block_count, threads_per_block, \
```
Is this still needed?

whchung replied:
This is one of the most important feature in this PR. Quite a few TF operators which depends on CUDA-specific `<<< >>>` kernel launch syntax would be changed to this macro.

Granted, I see several operators now moving toward using `CudaLaunchKernel` but still some of them employ `<<< >>>` therefore this macro is needed.

chsigg replied:
Ah, I see. Sounds good then. I will work towards using CudaLaunchKernel everywhere, but this is fine for now.

chsigg commented on tensorflow/core/util/gpu_kernel_helper.h:
```diff
@@ -13,34 +13,76 @@ See the License for the specific language governing permissions and
-#define CUDA_AXIS_KERNEL_LOOP(i, n, axis) \
-  for (int i : ::tensorflow::CudaGridRange##axis<int>(n))
+#if GOOGLE_CUDA
+#define gpuSuccess cudaSuccess
```
Are these still needed?

whchung replied:
These are needed so we have minimal impacts to the implementation of operators. Once set of `gpuXXX` macro would be built on both CUDA and ROCm.

chsigg commented on tensorflow/core/util/gpu_launch_config.h:
```diff
@@ -141,24 +141,41 @@ inline CudaLaunchConfig GetCudaLaunchConfig(int work_element_count,
+  const int physical_thread_count = std::min(
+      d.getNumGpuMultiProcessors() * d.maxGpuThreadsPerMultiProcessor(),
+      work_element_count);
+  thread_per_block = std::min(1024, d.maxGpuThreadsPerBlock());
```
Could you add a comment what the '1024' is about?

whchung replied:
Will do.

whchung replied:
Turns out the logic is just a clone of `GetGpuLaunchConfig(int, const Eigen::GpuDevice&)`. Added comments to address that.

chsigg commented on tensorflow/core/util/gpu_launch_config.h:
```diff
@@ -141,24 +141,41 @@ inline CudaLaunchConfig GetCudaLaunchConfig(int work_element_count,
+  thread_per_block = std::min(1024, d.maxGpuThreadsPerBlock());
+  block_count =
+      std::min(DivUp(physical_thread_count, thread_per_block),
+               d.getNumGpuMultiProcessors());
```
Is the one block per SM limit intentional?

whchung replied:
I'll need to check with HIP runtime team on this heuristic. It was given by them before we really have `hipOccupancyMaxPotentialBlockSize` implemented in HIP runtime.

whchung replied:
Turns out the logic is just a clone of `GetGpuLaunchConfig(int, const Eigen::GpuDevice&)` above which assumes the kernel be simple and memory-bounded.

whchung replied:
Added comment to reflect this.

chsigg commented on tensorflow/core/util/gpu_launch_config.h:
```diff
@@ -255,10 +295,27 @@ inline Cuda3DLaunchConfig GetCuda3DLaunchConfig(
       block_size_limit);
   CHECK_EQ(err, cudaSuccess);
+#elif TENSORFLOW_USE_ROCM
+  // XXX FIXME re-enable this after hipOccupancyMaxPotentialBlockSize is
```
'ROCM TODO'?

whchung replied:
Will do.

chsigg reviewed (commented):
I'm not familiar with how warps work on AMD. Can you run with warp size of 32 (@ half the occupancy), or do we _need_ to handle warp size 64 in the kernel implementations?

whchung commented on tensorflow/core/util/gpu_device_functions.h:
```diff
@@ -304,61 +332,93 @@ __device__ inline T CudaShuffleDownSync(unsigned mask, T value, unsigned delta,
 // convergence, see comment above for details.
 template <typename T>
-__device__ T CudaShuffleXorSync(unsigned mask, T value, int lane_mask,
+__device__ T GpuShuffleXorSync(unsigned mask, T value, int lane_mask,
```
@chsigg Please refer here to see ROCm-specific logic for `GpuShuffleXorSync` for various data types.

whchung commented:
@chsigg I should also mention that the purpose for this PR to is lay infrastructure for subsequent PRs for other TF operators which invokes custom HIP GPU kernels. Therefore the focus of this PR is not on the operators, but on the utility functions.

Take `GpuShuffleXorSync` as an example, you can find ROCm-specific logic in `gpu_device_functions.h`, but you won't (yet) find ROCm-specific logic in TF operators which depend on this utility function. I took this route to avoid breaking upstream implementation that much.

Shall this PR be merged, we'll then file additional PRs to enable those operators supported on ROCm incrementally like other PRs submitted and merged recently.

whchung commented:
@chsigg On existing AMD hardware warp (wavefront in AMD terminology) is 64.

chsigg commented:
> @chsigg On existing AMD hardware warp (wavefront in AMD terminology) is 64.

I'm aware of that. My question was:
> Can you run with warp size of 32 (@ half the occupancy), or do we need to handle warp size 64 in the kernel implementations?

whchung commented:
@chsigg We'll need to handle warp size of 64 in the kernel implementations. Therefore for some operators we do have to change the implementation of some kernels and they will be reflected in subsequent PRs.

whchung committed:
Address review comments and clang-format

whchung commented:
@chsigg I've addressed your review comments.

Also a ticket was raised on HIP repo some time ago to request the feature to properly estimate required launch config for HIP kernels: https://github.com/ROCm-Developer-Tools/HIP/issues/924 .

whchung commented:
@chsigg rebased the PR to resolve conflicts.

Also it seems now all `<<< >>>` has been removed. I'll change `GPU_LAUNCH_KERNEL` macro to adopt `CudaLaunchKernel`.

whchung committed:
Revise GPU_LAUNCH_KERNEL macro for CUDA path

For ROCm path, a HIP runtime API which has return status code is needed.

jlebar commented on tensorflow/core/util/gpu_kernel_helper.h:
```diff
@@ -13,34 +13,77 @@ See the License for the specific language governing permissions and
-#define CUDA_AXIS_KERNEL_LOOP(i, n, axis) \
-  for (int i : ::tensorflow::CudaGridRange##axis<int>(n))
+#if GOOGLE_CUDA
+#define gpuSuccess cudaSuccess
```
Why do we have both `gpuSuccess` and `GPUSUCCESS`?

whchung replied:
I'll remove unused `GPUSUCCESS`.

jlebar commented on tensorflow/core/util/gpu_kernel_helper.h:
```diff
@@ -13,34 +13,77 @@ See the License for the specific language governing permissions and
+#define GPUGETLASTERROR() cudaGetLastError()
+using gpuError_t = cudaError_t;
+#define GPUSUCCESSS cudaSuccess
+#elif TENSORFLOW_USE_ROCM
```
Please add some whitespace to make this easy to read; the `elif` gets lost.

whchung replied:
will do.

jlebar reviewed (commented):
Christian asked me to do a once-over on this.  LGTM with some changes to the macros to meet Google's style guide.

whchung commented on tensorflow/core/util/gpu_kernel_helper.h:
```diff
@@ -13,34 +13,77 @@ See the License for the specific language governing permissions and
-  for (int i : ::tensorflow::CudaGridRange##axis<int>(n))
+#if GOOGLE_CUDA
+#define gpuSuccess cudaSuccess
+#define GPUGETERRORSTRING(error) cudaGetErrorString(error)
```
after reviewing the patch, the macros aren't really necessary. I'll remove them.

whchung committed:
Address reviewer comments to remove dead / obsolete codes / macros

chsigg commented:
I've been trying to merge this PR, but the file and symbol renames are tricky, because we have a lot of internal references to them as well. It will be easier if I do the rename using internal tools. Then this PR will essentially add the macros and ROCm specific bits. Does that sound acceptable to you?

whchung commented:
@chsigg sure thing

whchung commented:
@chsigg Just want to understand where we are getting this PR merged?

chsigg commented:
I'm in the process of renaming the files still. This is not trivial because we have many dependencies on these internally, some of which I don't have access to, so I need to ask others to make the change.

To unblock you, I've submitted a change so that both set of files exist, one forwarding to the other. I've not started renaming any symbols yet, but we will be in a similar situation that this is best done from our side. My recommendation is to update this PR to use the old symbols but #include the new headers, and trim it down to the functional changes that you would like to add.

whchung commented:
> I'm in the process of renaming the files still. This is not trivial because we have many dependencies on these internally, some of which I don't have access to, so I need to ask others to make the change.
> 
> To unblock you, I've submitted a change so that both set of files exist, one forwarding to the other. I've not started renaming any symbols yet, but we will be in a similar situation that this is best done from our side. My recommendation is to update this PR to use the old symbols but #include the new headers, and trim it down to the functional changes that you would like to add.

got it. let me revise the PR and follow your suggestion.

/cc @jerryyin for awareness

jerryyin commented:
@whchung Thanks, next weeks merging should be fun ...

whchung committed:
Introduce ROCm-specific logic for GPU utility routines

whchung commented:
@chsigg revised this PR so preserve existing symbols, while introducing ROCm-specific logic.

whchung commented:
@chsigg a gentle ping. could you advise steps remaining to get the PR merged

chsigg commented on tensorflow/core/util/gpu_device_functions.h:
```diff
@@ -251,11 +261,20 @@ __device__ T CudaShuffleSync(unsigned mask, T value, int src_lane,
   asm volatile("mov.b64 %0, {%1,%2};" : "=d"(value) : "r"(lo), "r"(hi));
   return value;
+#elif TENSORFLOW_USE_ROCM
+  uint64_t tmp = static_cast<uint64_t>(value);
```
This doesn't look correct: the static_cast will do a conversion. You probably wanted to use reinterpret_cast, which is technically not legal either.

jlebar replied:
memcpy is the only legal way of doing type punning in C++.

chsigg commented on tensorflow/core/util/gpu_device_functions.h:
```diff
@@ -564,6 +638,33 @@ __device__ detail::ToTypeIfConvertible<U, T> CudaAtomicMax(T* ptr, U value) {
+#if TENSORFLOW_USE_ROCM
+
+/*
+ * CUDA runtime headers have the following defined
```
Shouldn't this comment be about atomicMin/Max rather than the ordinary min/max?

chsigg commented:
The PR description looks outdated. Would you mind to update it?

whchung commented:
Thanks for the feedback. Let me try revise the PR according to your comments.
